templates/common/_base/files/configure-ovs-network.yaml

```diff
@@ -364,76 +349,97 @@ contents:
       echo "OVS configuration successfully reverted"
     }
 
-    # Reloads NetworkManager if any configuration change was done
-    reload_nm() {
+    # Reloads NM NetworkManager profiles if any configuration change was done.
+    # Accepts a list of devices that should be re-connect after reload.
+    reload_profiles_nm() {
       if [ $nm_config_changed -eq 0 ]; then
         # no config was changed, no need to reload
         return
       fi
-      nm_config_changed=0
-      
-      echo "Reloading NetworkManager after configuration changes..."
 
-      # recycle network, so that existing profiles and priorities are re-evaluated
-      nmcli network off
-
-      # wait for no devices to show as connected
-      echo "Waiting for devices to disconnect..."
-      if ! timeout 60 bash -c "while nmcli -g DEVICE,STATE d | grep -v :unmanaged; do sleep 5; done"; then
-        echo "Warning: NetworkManager did not disconnect all devices"
-      fi
-      
-      # reload profiles and set networking back on
+      # reload profiles
       nmcli connection reload
-      nmcli network on
-      
-      # restart NetworkManager so that we can wait on `nm-online -s`
-      systemctl restart NetworkManager
 
-      # Wait until all profiles auto-connect
-      echo "Waiting for profiles to activate..."
-      if nm-online -s -t 60; then
-        echo "NetworkManager has activated all suitable profiles after reload"
-      else
-        echo "Warning: NetworkManager has not activated all suitable profiles after reload"
-      fi
+      # precautionary sleep of 10s (default timeout of NM to bring down devices)
+      sleep 10
+    
+      # After reload, devices that were already connected should connect again
+      # if any profile is available. If no profile is available, a device can
+      # remain disconnected and we have to explicitly connect it so that a
+      # profile is generated. This can happen for physical devices but should
+      # not happen for software devices as those always require a profile.
+      for dev in $@; do
+        # Only attempt to connect a disconnected device
+        local connected_state=$(nmcli -g GENERAL.STATE device show "$dev" || echo "")
+        if [[ "$connected_state" =~ "disconnected" ]]; then
+          # keep track if a profile by the same name as the device existed 
+          # before we attempt activation
+          local named_profile_existed=$([ -f "${NM_CONN_PATH}/${dev}" ] || [ -f "${NM_CONN_PATH}/${dev}.nmconnection" ] && echo "yes")
+          
+          for i in {1..10}; do
+              echo "Attempt $i to connect device $dev"
+              nmcli device connect "$dev" && break
```

Comment:
will this exit the script if it fails due to  set -eux?

Reply:
No, even though the command can fail, the logical operator is evaluated successfully which is what counts.

From https://wiki.bash-hackers.org/commands/builtin/set

-e | errexit | When set, the shell exits when a simple command in a command list exits non-zero (FALSE). This is not done in situations, where the exit code is already checked (if, while, until, \|\|, &&)
-- | -- | --




```
~/dev/ovn-org/libovsdb cache-indexes* ⇡ vagrant@devmaster
❯ bash -c "set -eux;nmcli d connect nonexistent; echo continue"
+ nmcli d connect nonexistent
Error: Device 'nonexistent' not found.

~/dev/ovn-org/libovsdb cache-indexes* ⇡ vagrant@devmaster
❯ bash -c "set -eux;nmcli d connect nonexistent && echo success; echo continue"
+ nmcli d connect nonexistent
Error: Device 'nonexistent' not found.
+ echo continue
continue

~/dev/ovn-org/libovsdb cache-indexes* ⇡ vagrant@devmaster
❯ bash -c "set -eux;nmcli d connect nonexistent && break; echo continue"
+ nmcli d connect nonexistent
Error: Device 'nonexistent' not found.
+ echo continue
continue
```